### ISSUE TITLE
[Bug]: Masks do not work in most ControlNet models

### ISSUE THREAD

**Issue by Vendaciousness** [Mikubill/sd-webui-controlnet#1638]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits of both this extension and the webui

### What happened?

I found this when I was pasting images into ControlNet v1.1.218, after copying them from the Preprocessor Preview.  Basically, if you paste an image into CN, then use the Preprocessor Preview, then draw on it with the mask tool, CN will ignore the image and only honor what was draw with the mask tool.

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/127359965/234f2cc7-72f9-42cc-b1f8-ed67753a0f26)

### Steps to reproduce the problem

1. Go to txt2img tab of Automatic1111
2. Enter a prompt and generate an image.
3. Right-click on the image and select "Copy Image"
4. Open ControlNet and paste the image into any module and enable it.
5. Click on Canny
6. Click on the "Run Preprocessor" icon next to the dropdown menu.
7. Draw on the image.
8. Hit generate, or "Run Preprocessor" again and notice the image is now gone.

### What should have happened?

Pre-Processor should have used the image, rather than just what is drawn.

### Commit where the problem happens

version: [v1.2.1]
python: 3.10.9  •  torch: 2.0.0+cu118
xformers: 0.0.17
gradio: 3.32.0
checkpoint: [0a03c339c7](https://google.com/search?q=0a03c339c7a6d721c38f449fc49c3ea34ee195f28feac4b8fa14990885de9e63)

### What browsers do you use to access the UI ?

Google Chrome

### Command Line Arguments

```Shell
--opt-sdp-attention --medvram --xformers
```

### List of enabled extensions

deforum | 780e8a3b (Sat Jun 3 14:02:11 2023)
780e8a3bf4385fe121b897826206476c0f16ef61

sd-webui-controlnet | 1b2aa4a9 (Tue Jun 6 02:23:03 2023)
1b2aa4a96e9241b96abfbce249284ed17fc98b11

### Console logs

```Shell
venv "G:\Code\stable-diffusion-webui\venv\Scripts\Python.exe"
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
Version: v1.2.1
Commit hash: 89f9faa63388756314e8a1d96cf86bf5e0663045
Installing requirements


Launching Web UI with arguments: --opt-sdp-attention --medvram --xformers
2023-06-13 21:37:47,940 - ControlNet - INFO - ControlNet v1.1.218
ControlNet preprocessor location: G:\Code\automatic\models
2023-06-13 21:37:48,036 - ControlNet - INFO - ControlNet v1.1.218
Loading weights [0a03c339c7] from G:\Code\stable-diffusion-webui\models\Stable-diffusion\divineelegancemix_V5.safetensors
*Deforum ControlNet support: enabled*
Loading general settings...
reading custom settings from deforum_settings.txt
Creating model from config: G:\Code\stable-diffusion-webui\configs\v1-inference.yaml
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 7.9s (import torch: 1.6s, import gradio: 1.2s, import ldm: 0.4s, other imports: 0.8s, list SD models: 0.4s, load scripts: 1.6s, create ui: 1.6s, gradio launch: 0.2s).
Loading VAE weights specified in settings: G:\Code\stable-diffusion-webui\models\VAE\vae-ft-mse-840000-ema-pruned.ckpt
Applying xformers cross attention optimization.
Textual inversion embeddings loaded(7): Asian-Less-Neg, awaitingtongue, corneo_mercy, CyberRealistic_Negative-neg, easynegative, FastNegativeEmbedding, FastNegativeEmbeddingStrong
Model loaded in 3.5s (load weights from disk: 0.5s, create model: 1.0s, apply weights to model: 0.9s, apply half(): 0.8s, load VAE: 0.2s).
100%|██████████████████████████████████████████| 35/35 [00:10<00:00,  3.33it/s]
Total progress: 100%|██████████████████████████| 35/35 [00:09<00:00,  3.88it/s]
2023-06-13 21:38:35,851 - ControlNet - INFO - Pixel Perfect Computation:40it/s]
2023-06-13 21:38:35,851 - ControlNet - INFO - resize_mode = ResizeMode.OUTER_FIT
2023-06-13 21:38:35,852 - ControlNet - INFO - raw_H = 512
2023-06-13 21:38:35,852 - ControlNet - INFO - raw_W = 1280
2023-06-13 21:38:35,853 - ControlNet - INFO - target_H = 512
2023-06-13 21:38:35,854 - ControlNet - INFO - target_W = 1280
2023-06-13 21:38:35,856 - ControlNet - INFO - estimation = 512.0
2023-06-13 21:38:35,857 - ControlNet - INFO - Preview Resolution = 512
2023-06-13 21:38:53,012 - ControlNet - INFO - Pixel Perfect Computation:
2023-06-13 21:38:53,012 - ControlNet - INFO - resize_mode = ResizeMode.OUTER_FIT
2023-06-13 21:38:53,014 - ControlNet - INFO - raw_H = 512
2023-06-13 21:38:53,014 - ControlNet - INFO - raw_W = 1280
2023-06-13 21:38:53,015 - ControlNet - INFO - target_H = 512
2023-06-13 21:38:53,015 - ControlNet - INFO - target_W = 1280
2023-06-13 21:38:53,015 - ControlNet - INFO - estimation = 512.0
2023-06-13 21:38:53,016 - ControlNet - INFO - Preview Resolution = 512
2023-06-13 21:38:58,968 - ControlNet - INFO - Loading model: control_v11p_sd15_canny [d14c016b]
2023-06-13 21:38:59,460 - ControlNet - INFO - Loaded state_dict from [G:\Code\stable-diffusion-webui\models\ControlNet\control_v11p_sd15_canny.pth]
2023-06-13 21:38:59,460 - ControlNet - INFO - Loading config: G:\Code\stable-diffusion-webui\models\ControlNet\control_v11p_sd15_canny.yaml
2023-06-13 21:39:01,834 - ControlNet - INFO - ControlNet model control_v11p_sd15_canny [d14c016b] loaded.
2023-06-13 21:39:01,961 - ControlNet - INFO - using mask as input
2023-06-13 21:39:01,964 - ControlNet - INFO - Loading preprocessor: none
2023-06-13 21:39:01,964 - ControlNet - INFO - Pixel Perfect Computation:
2023-06-13 21:39:01,966 - ControlNet - INFO - resize_mode = ResizeMode.OUTER_FIT
2023-06-13 21:39:01,968 - ControlNet - INFO - raw_H = 512
2023-06-13 21:39:01,970 - ControlNet - INFO - raw_W = 1280
2023-06-13 21:39:01,971 - ControlNet - INFO - target_H = 512
2023-06-13 21:39:01,973 - ControlNet - INFO - target_W = 1280
2023-06-13 21:39:01,975 - ControlNet - INFO - estimation = 512.0
2023-06-13 21:39:01,977 - ControlNet - INFO - preprocessor resolution = 512
100%|██████████████████████████████████████████| 35/35 [00:12<00:00,  2.87it/s]
Total progress: 100%|██████████████████████████| 35/35 [00:12<00:00,  2.78it/s]
Total progress: 100%|██████████████████████████| 35/35 [00:12<00:00,  2.92it/s]
```

### Additional information

Choosing the Inpaint preprocessors resulted in a traceback error, but I had to set the preprocessor to none or something first and couldn't duplicate the bug, so I'm just writing this one.

**Comment by huchenlei** [comment]
I think this is not a bug. The behaviour on such situation is not well defined in the first place. I think apply the mask on input image for non-inpaint modules might be the most useful behaviour. This will make some use cases don't need to go through 3rd party image editing tool, e.g. extract canny map for everything except a person's face. 

Current behaviour is also useful in some way. Let's say you want to scribble based on a red image. Current behaviour will be a good fit.

**Comment by Vendaciousness** [comment]
I see that using Inpaint is the only way to get a working mask with ControlNet. Perhaps you could disable the feature for the other models, since what it does now is not masking and serves no purpose. I would consider it a bug, since it is entirely reasonable for the user to assume the mask feature would word on all models.

And agreed, it would be useful to be able to mask out certain areas of an image you are working on, rather than have to edit the image in img2img Inpainting or with another program. 

Here is Wikipedia's definition of a mask:
In art, craft, and engineering, masking is the use of materials to protect areas from change, or to focus change on other areas. This can describe either the techniques and materials used to control the development of a work of art by protecting a desired area from change; or a phenomenon that (either intentionally or unintentionally) causes a sensation to be concealed from conscious attention.

Here is how the mask works normally in SD:
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/127359965/c26e7634-f44d-4df2-9aa7-46a3d5f047ec)
Here's how it works in ControlNet:
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/127359965/b97e3fdb-b5d9-4956-b867-80b3cec66511)

It's up to you, what you wish to do, but I would request you add masking functionality to all models.  That would be super useful, I think.

**Comment by aarongerber** [comment]
I wish we could mask each Controlnet separately. I may just want a face or body from the original image.

**Comment by huchenlei** [comment]
See feature proposal https://github.com/Mikubill/sd-webui-controlnet/issues/1736
